packages/next/src/client/components/router-reducer/reducers/server-action-reducer.ts

```diff
@@ -268,23 +273,17 @@ export function serverActionReducer(
             updatedTree: newTree,
             updatedCache: cache,
             includeNextUrl: Boolean(nextUrl),
+            canonicalUrl: mutable.canonicalUrl || state.canonicalUrl,
           })
 
           mutable.cache = cache
           mutable.prefetchCache = new Map()
         }
 
         mutable.patchedTree = newTree
-        mutable.canonicalUrl = href
```

Comment:
Not sure why this was here but it's unnecessary: `href` is `state.canonicalUrl` and doesn't need to be re-assigned every iteration